### PR TITLE
[Backport] ZGC class unloading

### PR DESCRIPTION
**Pull request by tanghaoth90**

### CODE REVIEW

**Review by mmyxym — COMMENTED** [review]
Change the title with [GC] 8212996

**Review by mmyxym — COMMENTED** [review]
not need to mentioned "co-authored"....

File: src/hotspot/share/code/nmethod.cpp [hunk 1]
```diff
@@ -762,6 +769,7 @@ nmethod::nmethod(
     code_buffer->copy_values_to(this);
     debug_info->copy_to(this);
     dependencies->copy_to(this);
+    clear_unloading_state();
```
Comment (mmyxym) [inline src/hotspot/share/code/nmethod.cpp:772]: guarded with ZGC?

File: src/hotspot/share/gc/z/zDriver.cpp [hunk 2]
```diff
@@ -301,6 +315,14 @@ void ZDriver::concurrent_process_non_strong_references() {
   ZHeap::heap()->process_non_strong_references();
 }
 
+void ZDriver::pause_unload_class() {
```
Comment (mmyxym) [inline src/hotspot/share/gc/z/zDriver.cpp:318]: fininsh_non_strong_refs is not paused. Shall we rename it to "unload_class"?

Comment (tanghaoth90) [reply src/hotspot/share/gc/z/zDriver.cpp]: renamed

**Review by mmyxym — APPROVED** [review]

File: src/hotspot/share/gc/z/zHeap.cpp [hunk 3]
```diff
@@ -350,6 +351,9 @@ bool ZHeap::mark_end() {
   // Process weak roots
   _weak_roots_processor.process_weak_roots();
```
Comment (linade) [inline src/hotspot/share/gc/z/zHeap.cpp:353]: line 341 MetaspaceGC::compute_new_size(); is moved to _unload.finish(). What's the reason we still leave it here?

**Review by linade — COMMENTED** [review]
src/hotspot/share/code/* might need more review. Other parts looks good (left two comments)

**Review by kuaiwei — APPROVED** [review]

**Pull request merged**